### PR TITLE
WIP: packaging, installation, and testing

### LINKED ISSUE #569: doctests inside datashader not being run anywhere?
E.g. I think `core.trimesh()`'s doctest (https://github.com/bokeh/datashader/blob/master/datashader/core.py#L201) fails. 

The tests seem to be run differently for travis, appveyor, python 2, python 3, conda package, but it looks like none of them runs doctests on datashader itself (something like `pytest datashader --doctest-modules ...` instead of `pytest datashader/tests --doctest-modules...`).

### PR DESCRIPTION
**Pull request by ceball**
Work in progress; this description will change and grow...

## Documented "admin" tasks in runnable way

Have used a cross-platform, make-like tool plus library of tasks common to pyviz projects to allow admin tasks to be run the same locally or on CI, and on different platforms. For more background, see https://github.com/pyviz/pyctdev/.

```
$ doit list
build_docs            build docs
develop_install       python develop install, with specified optional groups of dependencies (installed by conda only).
ecosystem_setup       Common conda setup (must be run in base env).
env_capture           Report all information required to recreate current conda environment
env_create            Create named environment if it doesn't already exist
env_export            Generate a pinned environment.yaml from specified env, filtering
list_envs             
miniconda_download    Download Miniconda3-latest
miniconda_install     Install Miniconda3-latest to location if not already present
package_build         Build and then test conda.recipe/ (or specified alternative).
package_upload        Upload package built from conda.recipe/ (or specified alternative).
test_all              Run "all" tests
test_all_quick        Run "all_quick" tests
test_examples         Run "examples" tests
test_examples_quick   Run "examples_quick" tests
test_flakes           Run "flakes" tests
test_unit             Run "unit" tests


$ doit info package_build

package_build

Build and then test conda.recipe/ (or specified alternative).

status     : run
 * The task has no dependencies.

params     : 
 - {'name': 'channel', 'long': 'channel', 'short': 'c', 'type': <class 'list'>, 'default': []}
 - {'name': 'recipe', 'long': 'recipe', 'type': <class 'str'>, 'default': ''}
 - {'name': 'test_python', 'long': 'test-python', 'type': <class 'list'>, 'default': []}
 - {'name': 'test_group', 'long': 'test-group', 'type': <class 'list'>, 'default': []}
 - {'name': 'test_requires', 'long': 'test-requires', 'type': <class 'list'>, 'default': []}
 - {'name': 'pkg_tests', 'long': 'pkg-tests', 'type': <class 'bool'>, 'default': True, 'inverse': 'no-pkg-tests'}
 - {'name': 'pin_deps_as_env', 'long': 'pin-deps-as-env', 'type': <class 'str'>, 'default': ''}
```

## added support for pip

`pip install datashader` passes same unit tests as conda package. It's not yet possible to install all dependencies required for examples using pip (#588).

## dependencies in one place, labelled for different purposes

Expressed the dependencies in one place, setup.py. The dependencies listed in setup.py are used:

  * for pip packages
  * for conda packages
  * for pip and conda developer installs
  * to generate examples environment.yaml

Dependencies are labelled as required, for tests, for examples, for building docs.

## testing of what users install

For pip (using tox) and conda (using conda-build), packages can be built, installed into an environment, and then tested. Tests can be over multiple environments (with different versions of python, different optional dependencies). Which tests to run is selectable.

E.g. `$ doit package_build --recipe base --test-python=py36 --test-python=py27 --test-group=unit`
  

## testing of what developers do

`doit develop_install` installs dependencies from setup.py using conda, and does a `pip install -e --no-deps .` on top. 

## conda packages

One noarch:python package is generated instead of (win/mac/linux, 32/64 bits, python 2.7/3.5/3.6). That package can subsequently be tested over multiple versions of python etc.

Also, reduced duplication in conda metadata. conda packages are templated, reading info from setup.py.

Dependencies can also be pinned exactly at package build time (e.g. to those in an existing environment), for an "application style" package if necessary.

~~Create datashader ("base"/"core") and datashader-examples ("examples") packages. The datashader-examples package has extra dependencies. The examples package is pinned to the datashader package.~~

~~Someone could install the datashader-examples package as an alternative to the examples environment.yml (allowing install into an existing environment rather than forcing an isolated one).~~

## datashader command

`datashader examples` (`datashader copy-examples` and `datashader fetch-data`); see https://github.com/pyviz/pvct/blob/master/README.md

## Environment.yaml generation

Can be generated with each package pinned by version and channel. Includes just direct dependencies. Can optionally include any groups in setup.py `extras_require`, e.g. `examples`.

## Unified how various tools are run/configured

All test commands are in tox.ini, and that is the *only* place they are recorded; tox.ini contains the definitions of the groups of tests and the commands that comprise each group. It also contains definitions of the various optional environments, if any (e.g. for testing with different groups of dependencies present). It also contains all the necessary tool config.

However, it's unfortunate that the tox.ini is a little cryptic (https://github.com/pyviz/pyct/issues/9). 

(Should fix #569.)

## Use external projects for testing

Replacing project and platform specific code here with external, cross project and platform tools. Check notebooks run without error, notebooks do not contain flakes: pytest plugin (nbsmoke). ~~Allows declaring which notebooks are e.g. "quick", which can then also later be selected to run using pytest's regular selection mechanism.~~

~~examples/conftest.py contains~~

tox.ini has a skip list, which can be overridden from the command line to run all notebooks. See doc/README.md.

## automatic versioning

Version via git tag. Version is stored only in one place in the repository (a git tag), and is written into packages via a .version (json) file.

Have used autover (via param).

### developer tests

run every commit, the same on travis (linux) and appveyor (windows). Testing typical developer setup (currently python 3). Dependencies come from same place (setup.py).

### packages

Every push of a pre-release tag, conda packages are built (built, installed to fresh env, tested) and then uploaded to anaconda.org pyviz/label/dev and test pypi (https://test.pypi.org/project/datashader/).

The same could happen for release tags, except upload to pyviz/label/main and pypi.

Means release procedure can be e.g. to keep doing: push pre-release tag, install package from testpypi / anaconda pyviz/label/dev, fix, push pre-release tag, ... until happy. Then push release tag. Or it can be done locally (same commands).

### Travis caching

Of miniconda env for the environment used to test typical developer setup every commit. Seems faster to download cache + update env + upload cache than to create env from scratch. But people will need to be aware of the caching, and know how to delete on travis if necessary (just a few clicks). 

-----

To do:
  * [x] (edit: for now I just run one notebook; see tox.ini nbsmoke_skip_run) Which subset of examples counts as 'quick'? And which self contained and/or small data? See #407. Maybe just don't run the examples on travis yet, and tackle this in the future?
  * [x] (edit: future work, travis/appeveyor declare channels at the top) Decide what to do about some per project config, e.g. recommended conda channels for devs, ...
  * [x] agree to adopt vX.Y.Z scheme for autover
  * [x] decide if `examples` symlink to `datashader/examples` is ok. Negative: windows git developers may need to set up their git/system to support it. Negative: you may not like "seeing notebooks twice when you are operating in git repo". Positive: no code needed to copy files at package time. Positive: no code needed to find examples that depends on whether development or package install.
  * [x] (edit: future work to tidy up if desired) decide what to do about python scripts in examples that weren't importable unless all example dependencies are installed (so e.g. pytest couldn't process the directory). Just did if __main__ hacks for now.
  * [ ] pillow was pinned to 4.3.0 on appveyor; need to test ds pil usage on windows. (where did problem show up - is it covered by unit tests + running the limited set of examples, which is what happens on appveyor?) 
  * [ ] now we dropped the examples package, should move topics examples and dependencies to pyviz (maybe future PR?)
  * [ ] now we dropped the examples package, how do users know to get holoviews - via docs? (holoviews is imported in user guide notebooks , but isn't a required dependency.)
  * [ ] how does taxi_prepreprocessing_example relate to download_sample_data.py? And was it intended for running as a script (because that's what it is now...)

To do for CB: 
  * [x] This PR includes some python 2 fixes, some of which were separately submitted. Will need to clean up.
  * [x] Check updated dependencies covers #549. Should also address #404.
  * [ ] Travis file modifications before merge, once testing on this branch is done.
  * [x] complete repo layout change: address #542 "ensure notebooks have links to files within the repository rather than direct links to files on github" while doing this
  * [x] restore nb link checking (and possibly other tests); https://github.com/pyviz/nbsmoke/issues/9
  * [x] markdown readme for pypi / long description; https://github.com/pyviz/pyct/issues/18
  * [x] grep for TODOs and check they're done
  * [x] update docs for installation (and building docs, since that can't be on travis)
  * [x] (edit make pyviz release notes on pyct wiki) add back some form of release notes (should be much shorter)

### CODE REVIEW

File: datashader/__init__.py [hunk 1]
```diff
@@ -1,6 +1,15 @@
...
+try:
+    from param.version import Version
+    __version__ = str(Version(fpath=__file__, archive_commit="$Format:%h$", reponame="datashader"))
+    del Version
+except:
+    import os, json
+    __version__ = json.load(open(os.path.join(os.path.split(__file__)[0],
+                                              '.version'), 'r'))['version_string']
+    del os
+    del json
```
Comment (jlstevens) [inline datashader/__init__.py]: I'm not sure why you want to use ``del``  here. Especially for ``os`` and ``json`` which are common imports from the standard lib...

Comment (jbednar) [reply datashader/__init__.py]: Seems like that's optional, but it's nice to tidy up the namespace and not leave incidentals like that hanging round to confuse tab complete and dir()...

**Comment by jbednar** [comment]
It might be nice to have a separate list of metadata types to preserve by default or on request, which so far would be things to do with Jupyter Dashboards or with RISE.

File: .travis.yml [hunk 2]
```diff
@@ -1,76 +1,91 @@
-# We deliberately don't use travis's language=python option because
-# we install miniconda and use conda to get python. Additionally, 
-# Travis's auto-install of python doesn't work on osx images (see
-# https://github.com/travis-ci/travis-ci/issues/4729).
+# TODO before merge:
+#
+#  * try switch to defaults
+#
+#  * restore previous sections
 language: generic
```
Comment (jbednar) [inline .travis.yml]: I'm going to go ahead and merge this, but I'll change "TODO before merge" to just "TODO".

**Pull request merged**